File: Go/work/src/benchmark/perfectnumber/perfectnumber.go
// perfectnumber.go
//
// vim: ft=go sw=4 ts=4
//
// Calculating perfect numbers in Go
//

package perfectnumber

import "sync"

// predicate to check for perfect numbers
func IsPerfect(n int) bool {
	sum := 0
	for i := 1; i < n; i++ {
		if n%i == 0 {
			sum += i
		}
	}
	return sum == n
}
func IsPerfectAsync(n int) <-chan bool {
	out := make(chan bool)

	go func() {
		defer close(out)
		out <- IsPerfect(n)
	}()

	return out
}

// list all perfect numbers for given limit
func PerfectNumbers(n int) []int {
	list := []int{}
	for i := 1; i < n; i++ {
		if IsPerfect(i) {
			list = append(list, i)
		}
	}
	return list
}
func PerfectNumbersAsync(n int) <-chan int {
	var wg sync.WaitGroup
	out := make(chan int)

	// start calculation
	wg.Add(n - 1)
	for i := 1; i < n; i++ {
		go func(x int) {
			defer wg.Done()
			if IsPerfect(x) {
				out <- x
			}
		}(i)
	}

	// close out channel when finished
	go func() {
		wg.Wait()
		close(out)
	}()

	return out
}

File: FSharp/Benchmark/Tasks/Perfectnumber.fs
// Calculate perfect numbers in F#

namespace Tasks

module Perfectnumber =

    // predicate to check whether a number is perfect
    let isPerfect n =
        let rec loop i sum =
            if n = i then
                sum = n
            else if n % i = 0 then
                loop (i + 1) (sum + i)
            else
                loop (i + 1) sum

        if n > 0 then
            loop 1 0
        else
            failwith "n must > 0"


    // generate a list of perfect numbers until given limit
    let perfectNumbers n =
        let rec loop i =
            if i = n then
                []
            else if isPerfect i then
                i :: loop (i + 1)
            else
                loop (i + 1)

        loop 1

    let rec range start stop =
        if start < stop then
            start :: (range (start + 1) stop)
        else
            []


    // async variant of perfect number generation
    let isPerfectAsync n = async { return isPerfect n, n }

    let perfectNumbersAsync n =
        range 1 n
        |> Seq.map isPerfectAsync
        |> Async.Parallel
        |> Async.RunSynchronously
        |> Seq.filter fst
        |> Seq.map snd
        |> Seq.toList
